File: lib/core/sub_app.dart
import 'package:flutter/material.dart';
import '../services/config_service.dart';
import '../services/share_content.dart';
import 'search_result.dart';

/// Abstract class that all sub-apps must implement.
/// Each sub-app is a self-contained module with its own UI and logic.
abstract class SubApp {
  /// Unique identifier for the app (used for storage and navigation)
  String get id;

  /// Display name shown in the launcher
  String get name;

  /// Icon displayed in the launcher grid
  IconData get icon;

  /// Primary theme color for the app
  Color get themeColor;

  /// Builds the main widget for this sub-app
  Widget build(BuildContext context);

  /// Called when the app is first initialized
  void onInit() {}

  /// Called when the app is paused (goes to background)
  void onPause() {}

  /// Called when the app is resumed (comes to foreground)
  void onResume() {}

  /// Called when the app is disposed
  void onDispose() {}

  // Config management methods

  /// Get a config value (checks user override, then default)
  String? getConfig(String key, {String? defaultValue}) {
    return ConfigService.instance.get(key, appId: id, defaultValue: defaultValue);
  }

  /// Set a user override for a config value
  Future<void> setConfig(String key, String value) async {
    await ConfigService.instance.set(key, value, appId: id);
  }

  /// Define a default config value (call in onInit)
  void defineConfig(String key, String defaultValue) {
    ConfigService.instance.setDefault(key, defaultValue, appId: id);
  }

  /// Reset a config value to its default
  Future<void> resetConfig(String key) async {
    await ConfigService.instance.reset(key, appId: id);
  }

  /// Check if a config key is using its default value
  bool isDefaultConfig(String key) {
    return ConfigService.instance.isDefault(key, appId: id);
  }

  /// Get all config values for this app
  Map<String, String> getAllConfigs() {
    return ConfigService.instance.getAll(appId: id);
  }

  // Share receiver methods

  /// Content types this app can receive (empty = can't receive)
  List<ShareContentType> get acceptedShareTypes => [];

  /// Called when content is shared to this app.
  /// Override this to handle incoming shared content.
  Future<void> onReceiveShare(ShareContent content) async {}

  // Search methods

  /// Whether this app supports search
  bool get supportsSearch => false;

  /// Search for content within this app
  /// Override this to implement app-specific search
  Future<List<SearchResult>> search(String query) async {
    return [];
  }

  /// Navigate to a specific search result within this app
  /// Override this to implement app-specific navigation
  void navigateToSearchResult(
      BuildContext context, SearchResult result) {}
}

File: playground/Playground/Playground/Sources/Core/SubApp.swift
import SwiftUI

/// Protocol that all sub-apps must implement
/// Defines the interface for modular apps within the Playground container
@MainActor
protocol SubApp: AnyObject, Identifiable, ObservableObject {
    /// Unique identifier for the app (e.g., "chat", "notes", "vocabulary")
    var id: String { get }

    /// Display name shown in launcher and UI
    var name: String { get }

    /// SF Symbol icon name for the app
    var iconName: String { get }

    /// Theme color for the app
    var themeColor: Color { get }

    /// Whether this app supports global search
    var supportsSearch: Bool { get }

    /// Whether this app supports sharing content
    var supportsSharing: Bool { get }

    /// Accepted content types for sharing (e.g., ["text/plain", "image/png"])
    var acceptedShareTypes: [String] { get }

    /// Build the main view for this app
    @ViewBuilder
    func buildView() -> AnyView

    /// Lifecycle: Called when app is first initialized
    func onInit() async

    /// Lifecycle: Called when app enters foreground
    func onResume() async

    /// Lifecycle: Called when app enters background
    func onPause() async

    /// Lifecycle: Called when app is disposed
    func onDispose() async

    /// Search within this app (if supportsSearch is true)
    func search(query: String) async -> [SearchResult]

    /// Navigate to a specific search result
    func navigateToSearchResult(result: SearchResult) async

    /// Handle shared content from other apps (if supportsSharing is true)
    func onReceiveShare(content: SharedContent) async
}

// Default implementations for optional functionality
extension SubApp {
    var supportsSearch: Bool { false }
    var supportsSharing: Bool { false }
    var acceptedShareTypes: [String] { [] }

    func onInit() async {}
    func onResume() async {}
    func onPause() async {}
    func onDispose() async {}

    func search(query: String) async -> [SearchResult] { [] }
    func navigateToSearchResult(result: SearchResult) async {}
    func onReceiveShare(content: SharedContent) async {}
}

// MARK: - Search Support

/// Represents a search result from a sub-app
struct SearchResult: Identifiable {
    let id: String
    let type: SearchResultType
    let appId: String
    let title: String
    let subtitle: String?
    let preview: String?
    let navigationData: [String: Any]
    let timestamp: Date?

    enum SearchResultType: String {
        case file
        case note
        case vocabularyWord
        case chat
        case chatMessage
        case course
        case activity
    }
}

// MARK: - Sharing Support

/// Represents content being shared between apps
struct SharedContent {
    let type: String // MIME type
    let data: Data
    let metadata: [String: Any]?
}
